### PR TITLE
KennaV2 - pckg and enhancements

### PR DESCRIPTION
**Pull request by orhovy**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/20618

## Required version of Demisto
4.5

## Does it break backward compatibility?
   - No

## Must have
- [x] Tests
- [ ] Documentation - soon
- [x] Code Review

## Technical writer review
Mention and link to the files that require a technical writer review.
- [ ] [YAML file](https://github.com/demisto/content/blob/d473387c88217851b6ebad6b30954e68d84d29b8/Integrations/KennaV2/KennaV2.yml)
- [x] [CHANGELOG](https://github.com/demisto/content/blob/d473387c88217851b6ebad6b30954e68d84d29b8/Integrations/KennaV2/CHANGELOG.md)
- [x] [Description file]
(https://github.com/demisto/content/blob/d473387c88217851b6ebad6b30954e68d84d29b8/Integrations/KennaV2/KennaV2_description.md)

### CODE REVIEW

File: Integrations/KennaV2/KennaV2.py [hunk 1]
```diff
@@ -0,0 +1,480 @@
+from json import JSONDecodeError
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: we already import that in commonserverpython

File: Integrations/KennaV2/KennaV2.py [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+        super().__init__(base_url=base_url, verify=verify, proxy=proxy, headers=header)
+
+    def http_request(self, message: str, suffix: str, params: Optional[dict] = None,
+                     data: Optional[dict] = None):
+        return super()._http_request(message, suffix, params=params, json_data=data)
+
+
+def test_module(client: Client, *_):
+    """
+    Performs basic get request to get item samples12
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: typo

File: Integrations/KennaV2/KennaV2.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+    """
+    res_vulnerabilities = client._http_request('GET', '/vulnerabilities')
+    res_assets = client._http_request('GET', '/assets')
+
+    if isinstance(res_vulnerabilities.get('vulnerabilities'), list) and isinstance(res_assets.get('assets'), list):
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: it's good to have 2 tests.
is it possible(and valid) to have only 1 asset/vuln? - please check and change if necessary.

File: Integrations/KennaV2/KennaV2.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        return 'ok', None, None
+    else:
+        raise Exception('Error occurred while trying to query the api.')
+
+
+# ----------------------------------------------- Auxiliary functions -------------------------------------------------
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: no need

File: Integrations/KennaV2/KennaV2.py [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+
+
+def connect_api(client: Client, message: str, suffix: str, params: Optional[dict] = None,
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: should be inside the client class

File: Integrations/KennaV2/KennaV2.py [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+                            inner_dict.update({inner_wanted_key: in_raw.get(inner_actual_key)})
+                        lst_inner.append(inner_dict)
+                    context.update({wanted_key[0]: lst_inner})
+            else:
+                context.update({wanted_key: raw.get(actual_key)})
+        context_list.append(context)
+    return context_list
+
+
+# ----------------------------------------- Commands Functions ---------------------------------------------------------
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: no need

File: Integrations/KennaV2/KennaV2.py [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+    # Check if response contains errors
+    if raw_response.get('errors'):
+        raise DemistoException(raw_response.get('errors'))
+    elif raw_response.get('error'):
+        raise DemistoException(raw_response.get('error'))
+    response_list = raw_response
+    return response_list
+
+
+def create_dict(raw_data: List[Dict[str, Any]], wanted_keys: List[Any], actual_keys: List[Any]) -> List[Dict[str, Any]]:
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: please change to a more indicative name, e.g:
parse_response/outputs, create_context_dict

Comment (bakatzir) [reply Integrations/KennaV2/KennaV2.py]: also, aux functions are usually above the client class

File: Integrations/KennaV2/KennaV2.py [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+    return context_list
+
+
+# ----------------------------------------- Commands Functions ---------------------------------------------------------
+
+
+def search_vulnerabilities(client: Client, args: dict) -> Tuple[str, Dict[str, Any], List[Dict[str, Any]]]:
+    """Search vulnerability command.
+    Args:
+        client:  BaseClient which connects to api
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: you give a Client obj, not BaseClient. please change in all occurrences.

File: Integrations/KennaV2/KennaV2.py [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+        Raw Data
+    """
+    url_suffix = '/vulnerabilities/search'
+    human_readable = []
+    context: Dict[str, Any] = {}
+    params = {
+        'id' + '[]': args.get('id'),
+        'top_priority' + '[]': args.get('top-priority'),
+        'min_risk_meter_score': args.get('min-score'),
+        'status' + '[]': args.get('status')
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: please use f notation

File: Integrations/KennaV2/KennaV2.py [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+                       'Status', 'Threat', 'TopPriority',
+                       ['ServiceTicket', 'DueDate', 'ExternalIdentifier', 'Status', 'TicketType']]
+        actual_keys = ['asset_id', ['connectors', 'connector_definition_name', 'id', 'name', 'vendor'], 'cve_id',
+                       'fix_id',
+                       'id', 'patch', 'lisk_meter_score',
+                       ['scanner_vulnerabilities', 'external_unique_id', 'open', 'port'],
+                       'score', 'severity', 'status', 'threat', 'top_priority',
+                       ['service_ticket', 'due_date', 'external_identifier', 'status', 'ticket_type']]
+
+        context_list = create_dict(vulnerability_list, wanted_keys, actual_keys)
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: is it a list or a dict? function name and var are misleading

File: Integrations/KennaV2/KennaV2.py [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+        Raw Data
+    """
+    url_suffix = '/connectors'
+    human_readable = []
+    context: Dict[str, Any] = {}
+    connectors = connect_api(client=client, message='GET', suffix=url_suffix).get('connectors')
+    if connectors:
+        wanted_keys = ['Host', 'Name', 'Running', 'ID']
+        actual_keys = ['host', 'name', 'running', 'id']
+        context_list = create_dict(connectors, wanted_keys, actual_keys)
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: same as in search_vuln, please change in all occurrences

File: Integrations/KennaV2/KennaV2.py [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+def run_connector(client: Client, args: dict) -> Tuple[str, Dict[str, Any], List[Dict[str, Any]]]:
+    """Run Connector command.
+    Args:
+        client:  BaseClient which connects to api
+        args: arguments for the request
+    Returns:
+        Success/ Failure , according to the response
+    """
+    args_id: str = str(args.get('id'))
+    url_suffix = '/connectors/' + args_id + '/run'
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f notation

File: Integrations/KennaV2/KennaV2.py [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+    run_response = connect_api(client=client, message='GET', suffix=url_suffix)
+    if run_response:
+        if run_response.get('success') == 'true':
+            return 'Connector ran successfully!', {}, []
+        else:
+            return 'Connector did not run successfully!', {}, []
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: please add to the msg the connector id, and `!` -> `.`

File: Integrations/KennaV2/KennaV2.py [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+    else:
+        return "error from response", {}, []
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: what is the error? can you try to extract it from the response?

File: Integrations/KennaV2/KennaV2.py [hunk 15]
```diff
@@ -0,0 +1,480 @@
...
+        Raw Data
+    """
+    human_readable_markdown = ''
+    url_suffix = '/fixes/search'
+    context: Dict[str, Any] = {}
+    params = {
+        'id' + '[]': args.get('id'),
+        'top_priority' + '[]': args.get('top-priority'),
+        'min_risk_meter_score': args.get('min-score'),
+        'status' + '[]': args.get('status'),
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f notation

File: Integrations/KennaV2/KennaV2.py [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+    """
+    url_suffix = '/assets/search'
+    human_readable = []
+    context: Dict[str, Any] = {}
+    params = {
+        'id' + '[]': args.get('id'),
+        'hostname' + '[]': args.get('hostname'),
+        'min_risk_meter_score': args.get('min-score'),
+        'ip_address' + '[]': args.get('ip-address'),
+        'tags' + '[]': args.get('tags')
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f

File: Integrations/KennaV2/KennaV2.py [hunk 17]
```diff
@@ -0,0 +1,480 @@
...
+        for lst in vulnerability_list:
+            human_readable.append({
+                'id': lst.get('id'),
+                'Name': lst.get('cve_id'),
+                'Score': lst.get('risk_meter_score')
+            })
+        context = {
+            'Kenna.Vulnerabilities(val.ID === obj.ID)': context_list
+        }
+        human_readable_markdown = tableToMarkdown('Kenna Vulnerabilities', human_readable)
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: we can order the columns in the HR, and remove nulls. please do that in all func occurrences.
thumb rule is to choose 4-6 headers for each function.

File: Integrations/KennaV2/KennaV2.py [hunk 18]
```diff
@@ -0,0 +1,480 @@
...
+    """
+    params_to_update: dict = {}
+    args_id = str(args.get('id'))
+    status = str(args.get('status'))
+    notes = str(args.get('notes'))
+    if notes:
+        params_to_update['vulnerability'].update({'notes': notes})
+    if status:
+        params_to_update['vulnerability'].update({'status': status})
+    url_suffix = '/vulnerabilities/' + args_id
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f

File: Integrations/KennaV2/KennaV2.py [hunk 19]
```diff
@@ -0,0 +1,480 @@
...
+    try:
+        connect_api(client=client, message='PUT', suffix=url_suffix, data=params_to_update)
+    except DemistoException as exp:
+        if type(exp.__context__) == JSONDecodeError:
+            return 'Asset ' + str(args_id) + ' was updated', {}, []
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f

File: Integrations/KennaV2/KennaV2.py [hunk 20]
```diff
@@ -0,0 +1,480 @@
...
+    """Add tags command.
+    Args:
+        client:  BaseClient which connects to api
+        args: arguments for the request
+    Returns:
+        Success/ Failure , according to the response
+    """
+    args_id = str(args.get('id'))
+    tags = str(args.get('tag'))
+    url_suffix = '/assets/' + args_id + '/tags'
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f

File: Integrations/KennaV2/KennaV2.py [hunk 21]
```diff
@@ -0,0 +1,480 @@
...
+        }
+    }
+    try:
+        connect_api(client=client, message='PUT', suffix=url_suffix, data=asset)
+    except DemistoException as exp:
+        if type(exp.__context__) == JSONDecodeError:
+            return 'Tag ' + tags + ' was added to asset ' + args_id, {}, []
+        else:
+            return 'Tag ' + tags + ' was not added to asset ' + args_id, {}, []
+    return 'error', {}, []
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: please check if you can add the error msg

File: Integrations/KennaV2/KennaV2.py [hunk 22]
```diff
@@ -0,0 +1,480 @@
...
+    asset = {
+        'asset': {
+            'tags': tags
+        }
+    }
+    try:
+        connect_api(client=client, message='PUT', suffix=url_suffix, data=asset)
+    except DemistoException as exp:
+        if type(exp.__context__) == JSONDecodeError:
+            return 'Tag ' + tags + ' was added to asset ' + args_id, {}, []
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f

File: Integrations/KennaV2/KennaV2.py [hunk 23]
```diff
@@ -0,0 +1,480 @@
...
+        else:
+            return 'Tag ' + tags + ' was not added to asset ' + args_id, {}, []
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: f

File: Integrations/KennaV2/KennaV2.py [hunk 24]
```diff
@@ -0,0 +1,480 @@
...
+        connect_api(client=client, message='DELETE', suffix=url_suffix, data=asset)
+    except DemistoException as exp:
+        if type(exp.__context__) == JSONDecodeError:
+            return 'Tag ' + str(tags) + ' was deleted to asset ' + args_id, {}, []
+        else:
+            return 'Tag ' + str(tags) + ' was not deleted to asset ' + args_id, {}, []
+    return 'error', {}, []
+
+
+# -------------------------------------------- Main Function -----------------------------------------------------------
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: no need

File: Integrations/KennaV2/KennaV2.py [hunk 25]
```diff
@@ -0,0 +1,480 @@
...
+        'kenna-get-asset-vulnerabilities': get_asset_vulnerabilities,
+        'kenna-add-tag': add_tags,
+        'kenna-delete-tag': delete_tags,
+    }
+
+    try:
+        if command in commands:
+            return_outputs(*commands[command](client, demisto.args()))
+        else:
+            raise NotImplementedError
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: please add a message as well.

File: Integrations/KennaV2/KennaV2.py [hunk 26]
```diff
@@ -0,0 +1,480 @@
...
+
+    except Exception as e:
+        return_error(f'Error from Example Integration {e}', e)
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.py]: Example -> KennaV2
e -> err

File: Integrations/KennaV2/KennaV2.yml [hunk 27]
```diff
@@ -0,0 +1,447 @@
...
+      type: Number
+    - contextPath: Kenna.Vulnerabilities.Connectors.DefinitionName
+      description: Connector definition name related to vulnerability
+      type: String
+    - contextPath: Kenna.Vulnerabilities.Connectors.ID
+      description: Connector ID related to vulnerability
+      type: Number
+    - contextPath: Kenna.Vulnerabilities.Connectors.Name
+      description: Connector name related to vulnerability
+      type: Unknown
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2.yml]: fix all outputs types, should not be unknown.

File: Integrations/KennaV2/CHANGELOG.md [hunk 28]
```diff
@@ -0,0 +1,2 @@
+## [Unreleased]
+-
```
Comment (bakatzir) [inline Integrations/KennaV2/CHANGELOG.md]: please add valuable RN

File: Integrations/KennaV2/KennaV2_description.md [hunk 29]
```diff
@@ -0,0 +1,2 @@
+Kenna is a Risk Intelligence & Vulnerability platform that enables InfoSec
+  teams to prioritize and remediate vulnerabilities faster.
```
Comment (bakatzir) [inline Integrations/KennaV2/KennaV2_description.md]: please write here the instance config:
https://github.com/demisto/content/blob/master/Integrations/Palo_Alto_Networks_WildFire_v2/Palo_Alto_Networks_WildFire_v2_description.md
also supports MD.

**Comment by kirbles19** [comment]
@orhovy @bakatzir Done.

File: Integrations/KennaV2/KennaV2.yml [hunk 30]
```diff
@@ -0,0 +1,514 @@
+category: Vulnerability Management
+commonfields:
+  id: KennaV2
+  version: -1
+configuration:
+- defaultvalue: https://api.kennasecurity.com
+  display: Server URL (e.g. https://192.168.0.1)
```
Comment (glicht) [inline Integrations/KennaV2/KennaV2.yml]: Change example
```suggestion
  display: Server URL (e.g.  https://api.kennasecurity.com)
```

**Comment by glicht** [comment]
@kirbles19 please review readme file

**Comment by kirbles19** [comment]
@orhovy @bakatzir README finished.

**Pull request merged**